src/google/protobuf/descriptor.proto

```diff
@@ -481,7 +490,11 @@ message FieldOptions {
   // For Google-internal migration only. Do not use.
   optional bool weak = 10 [default=false];
 
-
+  // Provides the ability to override the name of the property
+  // generated for this field. This is applied to all properties
+  // and methods to do with this field, including HasFoo, FooCount,
+  // FooList etc.
+  optional string csharp_property_name = 11;
```

Comment:
The problem with C# is that properties as well as classes are using Pascal-case naming style. Therefore, the risk that you will have collision with other entities is higher for C# than for other languages (e.g. java has camel case fields/getters/setters vs pascal class names) - that's why this option is useful.
We've had quite a discussion about which options are necessary and which ones are not (see comments on #2) and ended up with opinion that csharp_property_name is the one we need.